Support submodule manipulation in GraphModule

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52358 Support submodule manipulation in GraphModule**

Differential Revision: [D26759260](https://our.internmc.facebook.com/intern/diff/D26759260)

Commit:
Support submodule manipulation in GraphModule

[ghstack-poisoned]

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -188,10 +174,11 @@ def __init__(self, root: Union[torch.nn.Module, Dict[str, Any]], graph: Graph, c
         else:
             raise RuntimeError('Unsupported type ' + str(root) + ' passed for root!')
         self.graph = graph
+        graph._gm = self
```
I don't think it makes sense for `GraphModule` to be assigning private attributes on `Graph`. I have a few other questions as well:

1) What happens if `Graph` is already owned by another `GraphModule`?
2) Going off 1, should we make it so that `_gm` is `None` in that situation? I guess the semantics could be "_gm contains the single owning `GraphModule` for this graph, otherwise (0 or 2+ owners) it is `None)"

jamesr66a replied:
Also I think line 181 is unnecessary since `self.graph = graph` will call the `graph` setter below

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -276,6 +264,113 @@ def _gen_model_repr(module_name: str, module: torch.nn.Module) -> Optional[str]:
+    def has_submodule(self, target: str) -> bool:
+        """
+        Returns whether or not this GraphModule contains the submodule
+        given by ``str``.
```
We probably want to give examples of fully-qualified names in these docstrings. It's a prominent PyTorch concept but I'm not sure it's well known (it's not really encountered in normal PyTorch usage)

jamesr66a replied:
Probably want to say "given by ``target``"

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -276,6 +264,113 @@ def _gen_model_repr(module_name: str, module: torch.nn.Module) -> Optional[str]:
+        for item in atoms:
+            if not hasattr(mod, item):
+                return False
+            mod = getattr(mod, item)
```
`getattr(mod, item)` isn't guaranteed to always return an `nn.Module`. Probably want to have an `isinstance` check here and return false if that fails. By definition of this method, I think this should return `false` if passed `foo.bar.baz` and `foo.bar.baz` is a parameter/other attribute but not a module

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -276,6 +264,113 @@ def _gen_model_repr(module_name: str, module: torch.nn.Module) -> Optional[str]:
+
+        for item in prefix:
+
+            t = getattr(self, item, None)
```
Similarly, we should check if `t` is a non-`Module` in our traversal

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -276,6 +264,113 @@ def _gen_model_repr(module_name: str, module: torch.nn.Module) -> Optional[str]:
+            delattr(mod, submod)
+            return True
+
+        return delete_nested(self, target)
```
I don't think the logic here is quite correct: a `Module` could be used if any one of the following is True:
1) It has children that are used
2) Its forward is called directly via a `call_module` node
3) It has a non-Module attribute that is used from a `get_attr` node

I think this function should instead just delete the `Module` that was specified by `target`. Then optionally there can be a separate function that accounts for the 3 use scenarios above and GCs modules. (BTW I think that last part can be done by calling `GraphModule.named_modules()`, iterating over the `Nodes`, and accounting for what's used, rather than a recursive algorithm)

jamesr66a reviewed (commented):
Nice!

Commit:
Update on "Support submodule manipulation in GraphModule"




[ghstack-poisoned]

Commit:
Update on "Support submodule manipulation in GraphModule"




[ghstack-poisoned]

jamesr66a commented on test/test_fx.py:
```diff
@@ -1769,7 +1769,106 @@ def forward(self, x: torch.Tensor):
+
+
+        # Test that we added the "dropout" submodule
+        self.assertEqual(module_exists(a, "net_b.net_c.dropout"), True)
```
übernit: the `unittest` base class has special `assert` methods for checking if things are true, false, etc: https://docs.python.org/3/library/unittest.html#assert-methods

jamesr66a commented on torch/fx/graph.py:
```diff
@@ -121,14 +121,16 @@ def forward(self, x):
         self._insert = self._root.prepend
         self._len = 0
+        self._owners: int = 1 if owning_module else 0
+        self._gm = owning_module
```
I think the attribute should be named `_owning_module` as well

jamesr66a commented on torch/fx/graph.py:
```diff
@@ -121,14 +121,16 @@ def forward(self, x):
     For the semantics of operations represented in the ``Graph``, please see :class:`Node`.
     """
-    def __init__(self):
+    def __init__(self, owning_module: Optional["GraphModule"] = None):     # type: ignore[name-defined]
```
What's the typing comment here for?

ansley replied:
mypy error: `torch/fx/graph.py:124: error: Name 'GraphModule' is not defined  [name-defined]`. Note that this is with the forward reference `"GraphModule"`

jamesr66a replied:
Does it work if we gate an import of `GraphModule` with `if TYPE_CHECKING`? e.g. https://github.com/pytorch/pytorch/blob/64847c7f0b3559c6edc40f001619b80c7dc68ef7/torch/fx/node.py#L8

jamesr66a commented on torch/fx/graph.py:
```diff
@@ -713,19 +727,14 @@ def print_tabular(self):
               headers=['opcode', 'name', 'target', 'args', 'kwargs']))
 
-    def lint(self, root : Optional[torch.nn.Module] = None):
+    def lint(self):
```
This is technically BC-breaking, but we're in beta, so probably fine. The other alternative would be to accept the argument and print a deprecation warning

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -215,6 +202,8 @@ def graph(self, g) -> None:
         """
         self._graph = g
+        g._gm = self if not g._owners else None
+        g._owners += 1
```
It's still weird that `GraphModule` is reaching into private attributes of `Graph`. it's probably better to define a method/property on `Graph` like `register_owner(gm : GraphModule)` that encapsulates this logic. That can also be where the accounting for multiple owners happens

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -277,6 +266,165 @@ def _gen_model_repr(module_name: str, module: torch.nn.Module) -> Optional[str]:
+
+            return find_and_delete(getattr(o, prefix), path)
+
+        return find_and_delete(self, target)
```
This function doesn't need to be recursive anymore, right? Basically just iteratively traverse until the 2nd to last atom then delete the last atom from there

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -277,6 +266,165 @@ def _gen_model_repr(module_name: str, module: torch.nn.Module) -> Optional[str]:
+        used: List[str] = [name for node in self.graph.nodes 
+                           if node.op == "call_module" or node.op == "get_attr"
+                           for name in itertools.accumulate(node.target.split("."),
+                                                            lambda x, y: x + "." + y if y else x)]
```
This is pretty gnarly to read. Can we split this up into nested loops/if statements and have a clear comment about what the accumulate bit is doing (I believe it's registering all prefixes for every qualified name but it takes a while to parse that out

jamesr66a reviewed (commented):
Cool! Just a few more comments

Commit:
Update on "Support submodule manipulation in GraphModule"




[ghstack-poisoned]

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -277,6 +265,186 @@ def _gen_model_repr(module_name: str, module: torch.nn.Module) -> Optional[str]:
+            if not isinstance(self, torch.nn.Module):
+                return False
+
+        return True
```
This could also go on `nn.Module`; nothing here is specific to `GraphModule`.

I think another way to do this with the current Module API is `target in dict(mod.named_modules())` but that's O(number of children modules) and this is probably faster. Might be a good addition to `nn.Module`

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -277,6 +265,186 @@ def _gen_model_repr(module_name: str, module: torch.nn.Module) -> Optional[str]:
+        Deletes the given submodule from ``self``.
+
+        The module will not be deleted if ``target`` is not a valid
+        target.
```
Maybe want to mention `delete_all_unused_submodules` so the user knows about it if they're just reading this docstring

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -277,6 +265,186 @@ def _gen_model_repr(module_name: str, module: torch.nn.Module) -> Optional[str]:
+            if not hasattr(self, item):
+                return False
+
+            self = getattr(self, item)
```
Reassigning `self` is pretty surprising. I'd recommend doing something like `mod_itr = self` above this loop and reassinging that

jamesr66a reviewed (commented):
Nice! Just some small stuff

Commit:
Update on "Support submodule manipulation in GraphModule"




[ghstack-poisoned]

Commit:
Update on "Support submodule manipulation in GraphModule"




[ghstack-poisoned]

jamesr66a reviewed (commented):
This looks good from the FX perspective

@gchanan can you look at or direct someone to look at the changes to `nn.Module`? We found ourselves adding some helper methods to `GraphModule` but it seemed like they weren't specific to `GraphModule`, just the whole submodule hierarchy of PyTorch

jamesr66a commented on torch/nn/modules/module.py:
```diff
@@ -382,6 +382,141 @@ def add_module(self, name: str, module: Optional['Module']) -> None:
             raise KeyError("module name can't be empty string \"\"")
         self._modules[name] = module
 
+    def has_submodule(self, target: str) -> bool:
```
@jbschlosser can you take a look at these changes? We added these methods to `fx.GraphModule`, but they only manipulate the module hierarchy, so it seemed like they could be equally applicable to `nn.Module` generally. Do these look OK?

jbschlosser replied:
Are these methods being added to support fx-related stuff or for utility in the public API?

If the latter, I'm a bit ambivalent about how they fit with the traditional usage of `Module`s. In general, the recommendation is to define custom modules deriving from the base `torch.nn.Module` that implement the `forward()` function for their computation.

Since the `forward()` function can't be "rewritten", submodules cannot in general be added / removed dynamically without invalidating the `forward()` logic (at least for modules that aren't `nn.Sequential`-like containers AKA ones that don't simply iterate through submodules in their `forward()` implementation).

For example, I'm worried that someone will think it's perfectly reasonable to delete the final classification layer of their [ResNet module](https://github.com/pytorch/vision/blob/2e8c124ff7266437efd37c7a23684f64fe6904cd/torchvision/models/resnet.py#L244) to try to obtain image embeddings and then get confused when an error occurs during the forward pass.

Will discuss this more with the other nn folks tomorrow!

albanD replied:
I would agree with @jbschlosser  here that these API look very surprising for nn.Module.
Another thing I am weary about as well is that they allow users to create invalid nn.Modules very easily. Which would hint to the fact that they should not be public.

Also, for the namings issue with nesting, `.named_modules()` should give you all the info you need. In particular you can check the keys to know if a submodule exist in a nested way.

jamesr66a replied:
@albanD `.named_modules()` has performance issues in that a query against it is `O(N)` in the number of transitive modules. We're already seeing poor performance in internal ads workloads because of it. I agree that the submodule manipulation APIs here don't make sense, but it's pretty critical that we have a way to query for the existence of a submodule without blowing up the complexity of the program

jamesr66a replied:
That said, I'm okay with moving all of this back to `GraphModule` to unblock, but I thought I'd make you aware of the issue

albanD replied:
Yes thanks for sharing!
As Joel mentioned above, the getter sounds like it would be useful just for that reason so that one would be ok to add I think.
But the setter looks much more dangerous and out of place.

jamesr66a commented on torch/nn/modules/module.py:
```diff
@@ -382,6 +382,141 @@ def add_module(self, name: str, module: Optional['Module']) -> None:
+            if not isinstance(self, torch.nn.Module):
+                return False
+
+        return True
```
Actually, can we make this `get_submodule` and it returns the `Module` if it exists or `None` otherwise? The existence check can just be `get_submodule(target) is None` then

jbschlosser commented on torch/nn/modules/module.py:
```diff
@@ -382,6 +382,141 @@ def add_module(self, name: str, module: Optional['Module']) -> None:
+            submod = getattr(mod, item, None)
+
+            if submod is None:
+                submod = torch.nn.Module()
```
Adding a base `Module` here has some problems. It is generally intended that modules derive from `nn.Module` and implement the `forward()` function for their computation. For the base `Module` added here, there will be no `forward()` implemented, so the submodule won't be usable without some hackery.

```python
import torch
from torch import nn

class MyModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.add_submodule('sub.linear1', nn.Linear(4, 3))
        self.add_submodule('sub.linear2', nn.Linear(3, 2))
        self.add_submodule('sub.linear3', nn.Linear(2, 1))

    def forward(self, x):
        return self.sub(x)

m = MyModule()
print(m(torch.randn(10, 4)))  # throws NotImplementedError
```

Commit:
Update on "Support submodule manipulation in GraphModule"




[ghstack-poisoned]

albanD commented on torch/nn/modules/module.py:
```diff
@@ -382,6 +382,151 @@ def add_module(self, name: str, module: Optional['Module']) -> None:
+        Args:
+            target: The fully-qualified string name of the submodule
+                to look for. (See above example for how to specify a
+                fully-qualified string.)
```
nit you might want to specify that such `fully-qualitfied` name is the one that is used in the state_dict and `named_*` methods.

albanD commented on torch/nn/modules/module.py:
```diff
@@ -776,7 +921,7 @@ def _maybe_warn_non_full_backward_hook(self, inputs, result, grad_fn):
 
         # At this point we are sure that inputs and result are tuple of Tensors
-        out_grad_fn = set([r.grad_fn for r in result if r.grad_fn is not None])
+        out_grad_fn = {r.grad_fn for r in result if r.grad_fn is not None}
```
Why is a dict better than a set here?

ansley replied:
This is a set comprehension, so it's the same thing as what we had before. I changed it because `flake8` doesn't like the [set constructor with a list comprehension](https://pypi.org/project/flake8-comprehensions/) for creating the temporary list

jamesr66a replied:
Can we use the explicit set constructor with the bare generator expression:

```
set(r.grad_fn for r in result if r.grad_fn is not None)
```

IMO the fact that python uses the same tokens for both set and dict literals is confusing and making things more explicit can improve readability

albanD commented on torch/nn/modules/module.py:
```diff
@@ -382,6 +382,151 @@ def add_module(self, name: str, module: Optional['Module']) -> None:
+        module_atoms, param_name = atoms[:-1], atoms[-1]
+        mod: torch.nn.Module = self
+
+        for item in module_atoms:
```
This code is common to all 3 new functions here, can it be factored?

ansley replied:
Good point, thanks!

Commit:
Update on "Support submodule manipulation in GraphModule"


Differential Revision: [D26759260](https://our.internmc.facebook.com/intern/diff/D26759260)

[ghstack-poisoned]

Commit:
Update on "Support submodule manipulation in GraphModule"


Differential Revision: [D26759260](https://our.internmc.facebook.com/intern/diff/D26759260)

[ghstack-poisoned]

jamesr66a commented on torch/nn/modules/module.py:
```diff
@@ -382,6 +382,137 @@ def add_module(self, name: str, module: Optional['Module']) -> None:
+
+        for name, buffer in mod.named_buffers():
+            if name == buffer_name:
+                return buffer
```
This is O(N). Can we instead directly fetch the attribute `getattr(mod, buffer_name)` and assert that `buffer_name` is in `self._buffers`?

jamesr66a commented on test/test_fx.py:
```diff
@@ -1842,7 +1843,140 @@ def forward(self, x: torch.Tensor):
+                if not isinstance(gm, torch.nn.Module):
+                    return False
+
+            return True
```
The fact that this is a copy-paste of one of the APIs being tested indicates that this probably isn't meaningful. I think it should be sufficient just to test `get_submodule` and omit the `module_exist` checks here.

If you want to check that the new APIs work with the existing infrastructure in `nn.Module`, you probably want to replace `module_exists` with a check that ensures the qualified name appears in `named_modules`, `_modules`, etc

jamesr66a commented on torch/fx/graph.py:
```diff
@@ -480,6 +499,19 @@ def get_attr(self, qualified_name: str, type_expr: Optional[Any] = None) -> Node
+        if (self.owning_module and 
+                self.owning_module.get_submodule(qualified_name) is None
+                and self.owning_module.get_parameter(qualified_name) is None
+                and self.owning_module.get_buffer(qualified_name) is None):
```
I'm a bit worried about the perf implications of all these checks: internally they're redundantly doing the same traversal of the module hierarchy. Maybe that part can be split off into a `get_attr` on `nn.Module` which these other APIs call and which we can call here

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -309,6 +310,135 @@ def _gen_model_repr(module_name: str, module: torch.nn.Module) -> Optional[str]:
+                def join_fn(x: str, y: str) -> str:
+                    if y:
+                        return x + "." + y
+                    return x
```
code golf: `join_fn` in the `accumulate` call could be `lambda x, y: '.'.join([x, y] if y else [x])`

```
>>> x, y = 'x', None
>>> '.'.join([x, y] if y else [x])
'x'
>>> y = 'y'
>>> '.'.join([x, y] if y else [x])
'x.y'

```

ansley replied:
Very cool!

jamesr66a reviewed (approved):
Looks great! Some small stuff, but OK to land if the core `nn.Module` API stuff is good with @jbschlosser @albanD et al

albanD commented on torch/nn/modules/module.py:
```diff
@@ -382,6 +382,137 @@ def add_module(self, name: str, module: Optional['Module']) -> None:
+
+        param: torch.nn.Parameter = getattr(mod, param_name)
+
+        if not isinstance(param, torch.nn.Parameter):
```
I am not sure about this silent return of None here.
Maybe that should be an error? @jbschlosser

jbschlosser replied:
Hmm, I agree with @albanD, but to be even more Pythonic (in line with the "easier to ask for forgiveness than permission" idiom), I'd suggest throwing a descriptive error for both the non-existent and invalid type cases. Same for `get_buffer` and `get_submodule`.

Commit:
Update on "Support submodule manipulation in GraphModule"


Differential Revision: [D26759260](https://our.internmc.facebook.com/intern/diff/D26759260)

[ghstack-poisoned]

jbschlosser commented on torch/nn/modules/module.py:
```diff
@@ -382,6 +382,143 @@ def add_module(self, name: str, module: Optional['Module']) -> None:
+                fully-qualified string.)
+
+        Returns:
+            Optional[torch.Tensor]: The buffer referenced by
```
nit: Docs here still match the old behavior. Otherwise the changes lgtm!

Commit:
Update on "Support submodule manipulation in GraphModule"


Differential Revision: [D26759260](https://our.internmc.facebook.com/intern/diff/D26759260)

[ghstack-poisoned]

Commit:
Update on "Support submodule manipulation in GraphModule"


Differential Revision: [D26759260](https://our.internmc.facebook.com/intern/diff/D26759260)

[ghstack-poisoned]

Commit:
Update on "Support submodule manipulation in GraphModule"


Differential Revision: [D26759260](https://our.internmc.facebook.com/intern/diff/D26759260)

[ghstack-poisoned]

Commit:
Update on "Support submodule manipulation in GraphModule"


Differential Revision: [D26759260](https://our.internmc.facebook.com/intern/diff/D26759260)

[ghstack-poisoned]
